d2/src/test/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitorTest.java

```diff
@@ -0,0 +1,273 @@
+    UriPropertiesDualReadMonitorTestFixture fixture = new UriPropertiesDualReadMonitorTestFixture();
+    UriPropertiesDualReadMonitor monitor = fixture.getMonitor();
+
+    ScheduledExecutorService executor = fixture.getExecutor();
+
+    CountDownLatch done = fixture.getDoneSignal(oldLbProps.size() + newLbProps.size());
+    // randomly report properties from old and new Lbs but properties in same lb are reported in order
+    executor.execute(() -> runNext(fixture, oldLbProps, false));
+    executor.execute(() -> runNext(fixture, newLbProps, true));
+
+    try {
+      boolean completed = done.await(1000, TimeUnit.MILLISECONDS);
+      if (completed) {
+        // similarity eventually converge to 1
+        assertEquals((double) monitor.getMatchedUris() / (double) monitor.getTotalUris(), 1.0,
+            "Similarity score not 1. Match record: " + monitor.getMatchRecord(CLUSTER_1));
+      }
+      executor.shutdownNow();
+    } catch (InterruptedException e) {
+      Assert.fail("Tasks were interrupted");
+    }
+  }
+
+  // ensure properties for the same lb are reported in order
+  private void runNext(UriPropertiesDualReadMonitorTestFixture fixture, Queue<UriProperties> props, boolean fromNewLb) {
+    UriProperties p = props.poll();
+
+    if (p != null && !fixture._executor.isShutdown()) {
+      fixture._executor.execute(() -> {
+        reportAndVerifyState(fixture._monitor, p, fromNewLb);
+        runNext(fixture, props, fromNewLb);
+        fixture._doneSignal.countDown();
+      });
+    }
+  }
+
+  private void reportAndVerifyState(UriPropertiesDualReadMonitor monitor, UriProperties prop, boolean fromNewLb) {
+    monitor.reportData(CLUSTER_1, prop, fromNewLb);
+    // if reportData on the same cluster are NOT synchronized, the total uris and matched uris counts could become < 0
+    // or > 2.
+    // e.g: when total uris = 1, matched uris = 1, if reporting URI_PROPERTIES_URI_1_AND_2 for new and old Lbs are
+    // executed concurrently, both counts will decrement by 1 twice and become -1 first, and total uris will increment
+    // by 2 twice and become 4.
+    // We verify that doesn't happen no matter what order the data is reported between the old and new Lbs.
+    int totalUris = monitor.getTotalUris();
+    int matchedUris = monitor.getMatchedUris();
+    double similarity = (double) matchedUris / (double) totalUris;
+
+    assertTrue(totalUris >= 0 && totalUris <= 2);
+    assertTrue(matchedUris >= 0 && matchedUris <= 2);
+    assertTrue(similarity >= 0.0 && similarity <= 1.0, "Similarity score should be >= 0 and <= 1."
+        + " Match record: " + monitor.getMatchRecord(CLUSTER_1));
+  }
+
+  private ClusterMatchRecord createMatchRecord(UriProperties oldLb, UriProperties newLb, int uris, int matched) {
+    ClusterMatchRecord cluster = new ClusterMatchRecord();
+    cluster._oldLb = oldLb;
+    cluster._newLb = newLb;
+    cluster._uris = uris;
+    cluster._matched = matched;
+    return cluster;
+  }
```

Comment:
Maybe make that constructor `@VisibleForTesting`?

Reply:
yeah, updated